File: wpilibNewCommands/src/test/native/cpp/frc2/command/StartEndCommandTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <frc2/command/Commands.h>

#include "CommandTestBase.h"
#include "frc2/command/StartEndCommand.h"

using namespace frc2;
class StartEndCommandTest : public CommandTestBase {};

TEST_F(StartEndCommandTest, StartEndCommandSchedule) {
  CommandScheduler scheduler = GetScheduler();

  int counter = 0;

  auto command =
      cmd::StartEnd([&counter] { counter++; }, [&counter] { counter++; });

  scheduler.Schedule(command);
  scheduler.Run();
  scheduler.Run();
  scheduler.Cancel(command);

  EXPECT_EQ(2, counter);
}

File: commandsv2/src/test/java/edu/wpi/first/wpilibj2/command/StartEndCommandTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.wpilibj2.command;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.concurrent.atomic.AtomicBoolean;
import org.junit.jupiter.api.Test;

class StartEndCommandTest extends CommandTestBase {
  @Test
  void startEndCommandScheduleTest() {
    try (CommandScheduler scheduler = new CommandScheduler()) {
      AtomicBoolean cond1 = new AtomicBoolean();
      AtomicBoolean cond2 = new AtomicBoolean();

      StartEndCommand command = new StartEndCommand(() -> cond1.set(true), () -> cond2.set(true));

      scheduler.schedule(command);
      scheduler.run();

      assertTrue(scheduler.isScheduled(command));

      scheduler.cancel(command);

      assertFalse(scheduler.isScheduled(command));
      assertTrue(cond1.get());
      assertTrue(cond2.get());
    }
  }
}
